docs/tutorial/algorithm_interfaces.rst

```diff
@@ -1,2 +1,464 @@
+    :meth:`~.AlgorithmInterface.train_step` on steps along a temporal extent in
+    parallel. So the tensor shape is :math:`[TB,\ldots]`. This speeds up the
+    training forward step and the result will be reshaped to
+    :math:`[T,B,\ldots]` when fed to :meth:`~.AlgorithmInterface.calc_loss`.
+
+In most cases, we only need to override these interfaces while following their
+input/output signatures. Potentially we are able to accomplish many things through
+them.
+
+Argument conventions
+--------------------
+
+Here we assume that there is only a single algorithm in the pipeline. In fact, multiple
+algorithms could consist of an *algorithm hierarchy* so that the calling of
+their interfaces of the same name is performed in a nested way. We leave this
+advantaged usage to a later chapter :doc:`./customize_algorithms`.
+
+* ``inputs`` denotes the observations/states based on which
+  an algorithm's interface does computation at every step.
+* ``output`` is specially used to represent a step function's main output, e.g.,
+  action of an RL algorithm, latent embedding of an encoder algorithm, or goal
+  representation of a goal proposal algorithm.
+* ``state`` denotes episodic memory (e.g., RNN state) that is automatically reset
+  at the beginning of an episode. Any interface with ``state`` as its input must
+  return an updated ``state`` so that ALF will automatically feed the updated
+  one to the next calling of the same interface at the next time step.
+* ``info`` contains anything other than the interface's output and state, for
+  example, tensors for summary, action distributions for training, etc. Depending
+  on the context, it might have two alias names: ``train_info`` and ``rollout_info``.
+  ``train_info`` usually appears in the context where ``info`` is used for computing losses
+  and gradients. ``train_info`` is output by either an on-policy algorithm's
+  :meth:`~.AlgorithmInterface.rollout_step` or an off-policy algorithm's
+  :meth:`~.AlgorithmInterface.train_step`. ``rollout_info`` exists in the
+  context of off-policy training, where the ``info`` of :meth:`~.AlgorithmInterface.rollout_step`
+  is used for reference (e.g., it contains rollout action distributions for computing the
+  probability ratio in PPO).
+* ``loss_info`` should be an instance of :class:`~.data_structures.LossInfo`
+  that can be correctly processed by :meth:`~.Algorithm.update_with_gradient`.
+
+It might be a little difficult to digest all these pieces of information at one time
+for a new user. Don't worry! In the following we will revisit them one by one with
+examples.
+
+On-policy interfaces
+--------------------
+
+We modify the minimal working example by introducing a new actor-critic algorithm
+that wraps :class:`ActorCriticAlgorithm` and does an additional trivial job of
+inspecting tensor shapes. First we define a new data structure
+
+.. code-block:: python
+
+    MyACInfo = namedtuple("MyACInfo", ["ac", "zeros"])
```

Comment:
IMHO good implementation approaches are usually bottom-up and good tutorials goes top-down. I would suggest stating the goals first instead of the `MyACInfo` definition which is more like implementation details. This is because the user will most likely ask "why do I need `MyACInfo`" after reading this line.

The order that seems a bit more natural is

1. Explain what does "inspecting the shape" means
2. In order to do that, we need to inject code in `rollout_step`, `calc_loss` hooks
3. Explain why we need new data structure to pass information around

Reply:
Good point. I've added more explanations and re-arranged the presentation order.